File: 09. Graph/07. Disjoint Set Union (DSU)/DSU.js
class DisjointSet {
    constructor(n) {
        this.parent = Array.from({ length: n }, (_, i) => i); // Each element is its own parent
        this.rank = Array(n).fill(0); // Rank is initially 0
        this.size = Array(n).fill(1); // Size of each set is initially 1
    }

    // Find the representative (root) of the set containing x, with path compression
    find(x) {
        if (this.parent[x] !== x) {
            this.parent[x] = this.find(this.parent[x]); // Path compression
        }
        return this.parent[x];
    }

    // Union two sets containing x and y, using union by rank
    unionByRank(x, y) {
        let rootX = this.find(x);
        let rootY = this.find(y);

        if (rootX !== rootY) {
            if (this.rank[rootX] > this.rank[rootY]) {
                this.parent[rootY] = rootX; // Attach smaller tree under larger tree
            } else if (this.rank[rootX] < this.rank[rootY]) {
                this.parent[rootX] = rootY;
            } else {
                this.parent[rootY] = rootX; // If ranks are equal, attach under one and increment rank
                this.rank[rootX]++;
            }
        }
    }

    // Union two sets containing x and y, using union by size
    unionBySize(x, y) {
        let rootX = this.find(x);
        let rootY = this.find(y);

        if (rootX !== rootY) {
            if (this.size[rootX] > this.size[rootY]) {
                this.parent[rootY] = rootX;
                this.size[rootX] += this.size[rootY];
            } else {
                this.parent[rootX] = rootY;
                this.size[rootY] += this.size[rootX];
            }
        }
    }

    // Helper function to check if two elements are in the same set
    isConnected(x, y) {
        return this.find(x) === this.find(y);
    }
}


function main() {
    const n = 7; // Number of elements
    const ds = new DisjointSet(n);

    // Perform some union operations
    ds.unionByRank(1, 2);
    ds.unionByRank(2, 3);
    ds.unionBySize(4, 5);
    ds.unionBySize(6, 4);

    // Check connectivity
    console.log("Is 1 connected to 3?", ds.isConnected(1, 3) ? "Yes" : "No");
    console.log("Is 1 connected to 5?", ds.isConnected(1, 5) ? "Yes" : "No");

    // Find the root of a specific element
    console.log("Root of 2:", ds.find(2));
    console.log("Root of 4:", ds.find(4));
}

main();

File: 09. Graph/07. Disjoint Set Union (DSU)/DSU.go
package main

import (
	"fmt"
)

// ---------- Disjoint Set Structure ----------

type DisjointSet struct {
	parent []int
	rank   []int
	size   []int
}

// Constructor
func NewDisjointSet(n int) *DisjointSet {
	ds := &DisjointSet{
		parent: make([]int, n),
		rank:   make([]int, n),
		size:   make([]int, n),
	}

	for i := 0; i < n; i++ {
		ds.parent[i] = i
		ds.rank[i] = 0
		ds.size[i] = 1
	}

	return ds
}

// Find with Path Compression
func (ds *DisjointSet) Find(x int) int {
	if ds.parent[x] != x {
		ds.parent[x] = ds.Find(ds.parent[x])
	}
	return ds.parent[x]
}

// Union by Rank
func (ds *DisjointSet) UnionByRank(x, y int) {
	rootX := ds.Find(x)
	rootY := ds.Find(y)

	if rootX != rootY {
		if ds.rank[rootX] > ds.rank[rootY] {
			ds.parent[rootY] = rootX
		} else if ds.rank[rootX] < ds.rank[rootY] {
			ds.parent[rootX] = rootY
		} else {
			ds.parent[rootY] = rootX
			ds.rank[rootX]++
		}
	}
}

// Union by Size
func (ds *DisjointSet) UnionBySize(x, y int) {
	rootX := ds.Find(x)
	rootY := ds.Find(y)

	if rootX != rootY {
		if ds.size[rootX] > ds.size[rootY] {
			ds.parent[rootY] = rootX
			ds.size[rootX] += ds.size[rootY]
		} else {
			ds.parent[rootX] = rootY
			ds.size[rootY] += ds.size[rootX]
		}
	}
}

// Check if two elements are connected
func (ds *DisjointSet) IsConnected(x, y int) bool {
	return ds.Find(x) == ds.Find(y)
}

// ---------- Main ----------

func main() {
	n := 7
	ds := NewDisjointSet(n)

	// Perform some union operations
	ds.UnionByRank(1, 2)
	ds.UnionByRank(2, 3)
	ds.UnionBySize(4, 5)
	ds.UnionBySize(6, 4)

	// Check connectivity
	fmt.Println("Is 1 connected to 3?", ds.IsConnected(1, 3))
	fmt.Println("Is 1 connected to 5?", ds.IsConnected(1, 5))

	// Find roots
	fmt.Println("Root of 2:", ds.Find(2))
	fmt.Println("Root of 4:", ds.Find(4))
}